services/madoc-ts/src/utility/cache-helper.ts

```diff
@@ -0,0 +1,52 @@
+import cache from 'memory-cache';
+
+export async function cachePromise<T extends object>(
+  key: string,
+  getter: () => Promise<T>,
+  timeInMs: number
+): Promise<T> {
+  const resource = cache.get(key);
+  if (resource) {
+    return resource as T;
+  }
+
+  const result = await getter();
+  cache.put(key, result, timeInMs);
+  return result;
+}
+
+const swcPromises: Record<string, Promise<any> | null> = {};
+const DEFAULT_STALE_TIME = 1000 * 60 * 60 * 8; // 8 hours
+
+export async function cachePromiseSWR<T extends object | null>(
+  key: string,
+  getter: () => Promise<T>,
+  timeInMs: number,
+  staleTimeInMs: number = DEFAULT_STALE_TIME
+): Promise<T> {
+  if (swcPromises[key]) {
+    await swcPromises[key];
+  }
+  const resource = cache.get(key);
+  const staleResource = cache.get(`@stale/${key}`);
+  if (!resource) {
+    const promise = getter();
+    swcPromises[key] = promise;
+    promise.then(result => {
+      cache.put(key, result, timeInMs);
+      cache.put(`@stale/${key}`, result, staleTimeInMs);
```

Comment:
I checked the source code of `memory-cache`. It doesn't copy the `result`, so the extra memory to keep a stale copy is only the size of the key. The memory is released when you put the stale copy. It _will_ mean that for infrequent requests we will hold onto stale data for longer. I've set it to 8 hours for now. (The service will also restart the node process if max memory is reached).